### PR TITLE
Add GPT-J 6B support to the gpt-neo implementation

### LINKED ISSUE #11320: Irregular VRAM usage with gpt-neo inference with sequences longer than 250 tokens
## Environment info

- `transformers` version: 4.5.1 / HEAD
- Platform: Linux/Colab Pro
- Python version: 3.7
- PyTorch version (GPU?): 1.8.1 (CUDA 11.0)
- Tensorflow version (GPU?):
- Using GPU in script?: Yes, NVIDIA P100
- Using distributed or parallel set-up in script?:

### Who can help

@patil-suraj

## Information

Model I am using (Bert, XLNet ...): EleutherAI/gpt-neo-2.7B

The problem arises when using:
* [ ] the official example scripts: (give details below)
* [x] my own modified scripts: (give details below)

The tasks I am working on is:
* [ ] an official GLUE/SQUaD task: (give the name)
* [x] my own task or dataset: (give details below)

## To reproduce

1. Install transformers in a Colab Pro notebook
2. Run this script to log peak memory usage for inference with increasing sequence length: https://gist.github.com/finetuneanon/7ce0ed5090a27a383abffbbbc0433a29
3. Wait for it to crash with an OOM error in the attention matmul somewhere above sequence length 1850

Output:
```
1870 5436434432
ok 6535669248
1871 5436434432
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-2-f2aeed4489bd> in <module>()
     21     return_dict_in_generate=True,
     22     repetition_penalty=1.2,
---> 23     pad_token_id=tokenizer.eos_token_id
     24   )
     25   del ids

13 frames
/usr/local/lib/python3.7/dist-packages/transformers/models/gpt_neo/modeling_gpt_neo.py in _attn(self, query, key, value, causal_mask, masked_bias, attn_dropout, attention_mask, head_mask)
    238         key = key.to(torch.float32)
    239 
--> 240         attn_weights = torch.matmul(query, key.transpose(-1, -2))
    241         attn_weights = torch.where(causal_mask, attn_weights, masked_bias.to(attn_weights.dtype))
    242 

RuntimeError: CUDA out of memory. Tried to allocate 4.59 GiB (GPU 0; 15.90 GiB total capacity; 9.75 GiB already allocated; 4.60 GiB free; 10.42 GiB reserved in total by PyTorch)
```

The full output can be found here: https://gist.github.com/finetuneanon/c7292ea676f57f5bb63803685d80bf5b

The output has the format:
```
sequence_length occupied_cuda_memory_before_inference
ok peak_occupied_cuda_memory_during_inference
```

Doing inference with real text has the same issue.

## Expected behavior

I expected memory usage to increase steadily instead of jumping around wildly, but I am not sure if this might actually be the correct behaviour. If it is correct, reliably doing inference on long sequences on 16GB of VRAM seems to be impossible, but sometimes it works.

I have also plotted the peak memory allocation during inference:

cudaoom

The green line is peak memory allocation, the brown line is the amount of memory in use before running inference.

### PR DESCRIPTION
**Pull request by finetunej**
# What does this PR do?

This PR mainly adds support for the GPT-J 6B model. A [conversion script](https://gist.github.com/finetuneanon/ee196c6cd16af1de4ca444862414683a) and [config.json](https://gist.github.com/finetuneanon/a55bdb3f5881e361faef0e96e1d41f09) for the slim checkpoint are also available.

It also addresses the local attention issue from #11320 in the same way as PR #11630 and works around an issue with torch.multinomial that will allow zero probability tokens to be chosen when sampling from an fp16 model.

Special thanks to the great folks of the EleutherAI discord, who helped me debug the RoPE implementation and to @kurumuz (NovelAI) who worked on this as well.

Fixes #11320

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/models/gpt_neo/modeling_gpt_neo.py [hunk 1]
```diff
@@ -371,115 +337,15 @@ def forward(
         return outputs  # a, present, (attentions)
 
 
-class GPTNeoLocalSelfAttention(nn.Module, GPTNeoAttentionMixin):
```
Comment (patrickvonplaten) [inline src/transformers/models/gpt_neo/modeling_gpt_neo.py:374]: Why delete all this?

Comment (finetuneanon) [reply src/transformers/models/gpt_neo/modeling_gpt_neo.py:374]: The way local attention was implemented caused very high VRAM use with certain, especially prime, sequence lengths as described in #11320:

patched

The green line is the original implementation's VRAM use in GB over sequence length. Padding inputs to multiples of the local attention window sizes mitigates this, but the code is still hard to follow. The way local attention is implemented in this PR follows PR #11630. Instead of the very complex folding approach, it is calculated like global attention with an adjusted mask, which masks out attention values beyond the local attention window. This is much simpler, easier to understand and does not seem to have any noticable negative performance impact. It also fixes the issue with high VRAM usage.

Comment (patrickvonplaten) [reply src/transformers/models/gpt_neo/modeling_gpt_neo.py:374]: I see, it's mentioned in the other PR! Think we can go along with this after having done some memory & inference benchmarks

Comment (finetuneanon) [reply src/transformers/models/gpt_neo/modeling_gpt_neo.py:374]: I'm currently running benchmarks of both implementations. Will report back once they finish.

Comment (finetuneanon) [reply src/transformers/models/gpt_neo/modeling_gpt_neo.py:374]: You can find the full results in these notebooks:

[localattention_4.6.1.ipynb](https://github.com/finetuneanon/misc/blob/main/transformers_benchmark/localattention_4.6.1.ipynb)
[localattention_gpt-j.ipynb](https://github.com/finetuneanon/misc/blob/main/transformers_benchmark/localattention_gpt-j.ipynb)

My implementation of local attention seems to be significantly faster. It uses slightly more memory for sequence lengths that are divisible by the window size, but exhibits no sudden jumps in VRAM usage like the current implementation.

4.6.1, half precision:
```
torch.float16 cuda:0
seq_len	max_len	runtime	memory
64	104	1.2820s	5.059GB
127	167	1.4140s	5.090GB
251	291	1.8570s	5.181GB
256	296	1.8761s	5.180GB
383	423	2.8156s	6.994GB
509	549	3.3879s	7.646GB
512	552	2.7486s	5.317GB
769	809	4.6178s	9.000GB
1021	1061	5.7816s	10.304GB
1024	1064	4.4912s	5.590GB
1280	1320	5.3699s	5.735GB
1283	1323	7.0542s	11.664GB
1536	1576	6.2600s	5.882GB
1543	1583	8.2814s	13.013GB
1792	1832	7.1836s	6.023GB
1801	1841	9.5767s	14.350GB
[OOM error]
```

gpt-j branch, half precision:
```
torch.float16 cuda:0
seq_len	max_len	runtime	memory
64	104	0.9781s	5.186GB
127	167	1.0396s	5.225GB
251	291	1.1590s	5.310GB
256	296	1.1628s	5.311GB
383	423	1.3393s	5.401GB
509	549	1.4969s	5.490GB
512	552	1.4975s	5.493GB
769	809	1.8752s	5.683GB
1021	1061	2.1957s	5.861GB
1024	1064	2.1975s	5.861GB
1280	1320	2.5928s	6.051GB
1283	1323	2.6369s	6.053GB
1536	1576	2.9966s	6.243GB
1543	1583	3.0621s	6.249GB
1792	1832	3.3844s	6.422GB
1801	1841	3.4636s	6.426GB
2039	2049	2.6127s	6.597GB
2048	2049	2.2140s	6.458GB
```

**Comment by StellaAthena** [comment]
Just as a note, we have a PyTorch checkpoint for GPT-J that we will be ready to upload once this PR goes through.

**Comment by finetuneanon** [comment]
The conversion script in the top post generates a single file checkpoint, but for models of this size, I've found split up checkpoints usually more efficient to load and handle. Such split up checkpoints can be generated using [this conversion script](https://gist.github.com/finetuneanon/7dd417a31338a63f219a49702e0550db) and loaded as follows:

```python
import collections.abc

class Checkpoint(collections.abc.MutableMapping):
    def __init__(self, chkpt_dir):
        self.checkpoint = torch.load(f"{chkpt_dir}/m.pt")
    def __len__(self):
        return len(self.checkpoint)
    def __getitem__(self, key):
        return torch.load(self.checkpoint[key])
    def __setitem__(self, key, value):
        return
    def __delitem__(self, key, value):
        return
    def keys(self):
        return self.checkpoint.keys()
    def __iter__(self):
        for key in self.checkpoint:
            yield (key, self.__getitem__(key))
    def __copy__(self):
        return Checkpoint()
    def copy(self):
        return Checkpoint()

from transformers import GPTNeoForCausalLM, AutoConfig
config = AutoConfig.from_pretrained(model_name)
model = GPTNeoForCausalLM.from_pretrained(pretrained_model_name_or_path=None, config=config, state_dict=Checkpoint("checkpoint"))
```

Having a more integrated or better specifid way of loading them would be helpful, but I'm not sure what the best place for that would be.

**Comment by finetuneanon** [comment]
I noticed that there was a typo in the config file, which caused it to be invalid JSON. It's fixed now.

**Comment by finetuneanon** [comment]
Also ran some evaluations using the [eval harness](https://github.com/EleutherAI/lm-evaluation-harness) on the ported model now:

|   Task   |    Metric     |Value |
|----------|---------------|-----:|
|lambada   |ppl            |4.1060|
|          |ppl_stderr     |0.0886|
|          |acc            |0.6833|
|          |acc_stderr     |0.0065|
|winogrande|acc            |0.6480|
|          |acc_stderr     |0.0134|
|piqa      |acc            |0.7541|
|          |acc_stderr     |0.0100|
|          |acc_norm       |0.7612|
|          |acc_norm_stderr|0.0099|
|hellaswag |acc            |0.4895|
|          |acc_stderr     |0.0050|
|          |acc_norm       |0.6614|
|          |acc_norm_stderr|0.0047|

**Comment by StellaAthena** [comment]
The eval numbers are a little shy of what we have for the Jax model, but close enough that FP rounding could plausibly explain the difference:

Lambada: 3.99 ppl, 0.697 acc
Winogrande: 0.653
PiQA: 0.765
HellaSwag: 0.661

**Comment by finetuneanon** [comment]
It should also be noted that my results were with fp16. It should be easy enough to modify the conversion script to cast to fp32 (just replace `half()` with `float()`), which might give results closer to the original evaluation, but I don't currently have access to hardwhere which could run the model in fp32 at a reasonable speed to do an evaluation on it.

**Comment by StellaAthena** [comment]
> It should also be noted that my results were with fp16. It should be easy enough to modify the conversion script to cast to fp32 (just replace `half()` with `float()`), which might give results closer to the original evaluation, but I don't currently have access to hardwhere which could run the model in fp32 at a reasonable speed to do an evaluation on it.

Oh yes, I concur. That wasn’t meant as a detraction at all. I’m not sure if EAI had enough free GPUs, but I can look at getting the evals run at full precision later this week.

**Comment by shamoons** [comment]
Just curious - how long before new models are merged to the repo, generally speaking? And how long until it's available in the hosted inference API?

**Comment by patil-suraj** [comment]
Hi @finetuneanon 

Amazing, thanks a lot for porting `GPT-J` so quickly!

The changes to local attention look good to me. But would be nice to split the PR into two 
1. simplify local attention
2. and add GPT-J in a new model file.

While supporting `GPT-J` in the `GPTNeo` model is totally doable, we would like to avoid that. The overall philosophy is to combine more than one model only if the forward pass is exactly similar or requires some really minor changes. If I understand it correctly, here are the differences between `GPT-J` and `GPTNeo` :
- `GPT-J` uses rotary embeddings
- It scales attention weights
- no bias in the attention projection layer (the `out_proj` layer in attention class)
- does not use layer_norm before the feed forward layer (`mlp`)
- no residual connection between `hidden_states` and `attention_output` , just one residual connection which is added to `attention + mlp(hiddn)`
- uses bias in the output projection layer
- does not ties word embeddings with output layer

The current PR handles this using the `config.jax` argument, but it’s a bit confusing, and generally, one config param should not control these many changes. So if we really decide to support this in `GPTNeo` we would probably end-up with different config params like `attention_projection_bias`, `output_bias`, `attention_residual`,  `scale_attention`. So it’s cleaner IMO to add a new model for this.

Also, `Transformers` isn’t really a modular toolkit for building different models, The goal is to keep every model file responsible for one model so it becomes easier for everyone to understand it and modify it according to their needs.  And also makes it easy for us to maintain these different models.
cc @LysandreJik , @sgugger , @patrickvonplaten

Happy to help in any way to add the model :)

**Comment by finetuneanon** [comment]
To be quite honest, I think reading a couple of if branches makes the differences between models much clearer than having to compare two completely different model classes with small differences. You mention that transformers is not intended to be a modular framework, so there should be no issue with controlling these changes through a single configuration variable, although perhaps the naming of `jax` is not optimal. I would be open to changing this to e.g. `gptj`. Splitting it up into multiple options would only make sense to actually turn it into a modular framework.

**Comment by sgugger** [comment]
Let's agree to disagree: this is one of the core principle of the Transformers library, explicitly stated in our [philosophy](https://huggingface.co/transformers/philosophy.html). We've been proceeding like this since the beginning, and while we definitely understand where you're coming from, this is a defining principle of our library which we are not eager to change as it has been validated both by [surveys](https://discuss.huggingface.co/t/transformers-huge-community-feedback/120) and by community feedback.

Unfortunately, we will have to insist on GPT-J following the same approach as the rest of the models - for philosophy, maintenance and coherence's sake. Let us know if you would like us to take over, we are happy to! Thank you for your understanding.

**Comment by finetuneanon** [comment]
Yes, please take over in that case.

**Comment by StellaAthena** [comment]
> Let's agree to disagree: this is one of the core principle of the Transformers library, explicitly stated in our [philosophy](https://huggingface.co/transformers/philosophy.html). We've been proceeding like this since the beginning, and while we definitely understand where you're coming from, this is a defining principle of our library which we are not eager to change as it has been validated both by [surveys](https://discuss.huggingface.co/t/transformers-huge-community-feedback/120) and by community feedback.
> 
> Unfortunately, we will have to insist on GPT-J following the same approach as the rest of the models - for philosophy, maintenance and coherence's sake. Let us know if you would like us to take over, we are happy to! Thank you for your understanding.

If the current changes were to be submitted as a new model, instead of a modification to GPT-Neo, would there be any significant further changes to be made?

**Comment by sgugger** [comment]
> If the current changes were to be submitted as a new model, instead of a modification to GPT-Neo, would there be any significant further changes to be made?

The new modeling file is the main thing, we have a [template to add new models](https://github.com/huggingface/transformers/tree/master/templates/adding_a_new_model) that should take care of everything else.

**Comment by StellaAthena** [comment]
I have opened a PR that attempts to refit this into HF's paradigm. I recommend closing this PR and directing discussion to #12243 12243

**Comment by patil-suraj** [comment]
As discussed above, this PR will be split into two
- GPT-J (which Stella has already started)
- simplifying GPTNeo local attention 

Closing this PR now.
